### PR TITLE
 Replace old model to shared model in synchronizer

### PR DESCRIPTION
**Pull request by luckychess**
Signed-off-by: luckychess <toobwn@gmail.com>

### Description of the Change

Synchronizer: replace old model to shared model.

### Benefits

Old model is a subject to remove so we are one step closer to success now.

### Possible Drawbacks 
None.

### Usage Examples or Tests *[optional]*
Tests were changed accordingly.

### CODE REVIEW

File: irohad/synchronizer/impl/synchronizer_impl.cpp [hunk 1]
```diff
@@ -61,21 +63,18 @@ namespace iroha {
...
 
-        auto single_commit = rxcpp::observable<>::just(old_commit_message);
+        auto single_commit = rxcpp::observable<>::just(commit_message);
 
         notifier_.get_subscriber().on_next(single_commit);
       } else {
         // Block can't be applied to current storage
         // Download all missing blocks
-        for (auto signature : old_commit_message.sigs) {
+        for (auto signature : commit_message->signatures()) {
```
Comment (Warchant) [inline irohad/synchronizer/impl/synchronizer_impl.cpp]: const auto& signature

File: irohad/synchronizer/impl/synchronizer_impl.cpp [hunk 2]
```diff
@@ -37,16 +37,18 @@ namespace iroha {
           blockLoader_(std::move(blockLoader)) {
       log_ = logger::log("synchronizer");
       consensus_gate->on_commit().subscribe(
-          subscription_,
-          [&](model::Block block) { this->process_commit(block); });
+          subscription_, [&](model::Block block) {
```
Comment (Warchant) [inline irohad/synchronizer/impl/synchronizer_impl.cpp]: is it possible to to pass Block here by const ref?

File: irohad/synchronizer/impl/synchronizer_impl.cpp [hunk 3]
```diff
@@ -88,28 +87,21 @@ namespace iroha {
...
-
-            // TODO: 07-03-2018 Alexey Chernyshov remove this after relocation to
-            // shared_model https://soramitsu.atlassian.net/browse/IR-903
-            notifier_.get_subscriber().on_next(chain.map([](auto block) {
-              std::unique_ptr<iroha::model::Block> old_block(
-                  block->makeOldModel());
-              return *old_block;
-            }));
+            notifier_.get_subscriber().on_next(
+                chain.map([](auto block) { return block; }));
```
Comment (Warchant) [inline irohad/synchronizer/impl/synchronizer_impl.cpp]: is it possible to pass block by const ref?

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 4]
```diff
@@ -79,20 +69,17 @@ TEST_F(SynchronizerTest, ValidWhenInitialized) {
 
 TEST_F(SynchronizerTest, ValidWhenSingleCommitSynchronized) {
   // commit from consensus => chain validation passed => commit successful
-  Block test_block;
-  test_block.height = 5;
-  test_block.hash = iroha::hash(test_block);
+  std::shared_ptr<shared_model::interface::Block> test_block =
+      std::make_shared<shared_model::proto::Block>(
+          TestBlockBuilder().height(5).build());
```
Comment (Warchant) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: do not write like that. 

Simplify:
```C++
auto block = TestBlockBuilder().height(5).build();
auto test_block = std::make_shared<shared_model::proto::Block>(block);
```

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 5]
```diff
@@ -104,11 +91,11 @@ TEST_F(SynchronizerTest, ValidWhenSingleCommitSynchronized) {
 
   auto wrapper =
       make_test_subscriber<CallExact>(synchronizer->on_commit_chain(), 1);
-  wrapper.subscribe([&test_block](auto commit) {
+  wrapper.subscribe([test_block](auto commit) {
```
Comment (Warchant) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: is it possible to pass by const ref?

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 6]
```diff
@@ -104,11 +91,11 @@ TEST_F(SynchronizerTest, ValidWhenSingleCommitSynchronized) {
...
     auto block_wrapper = make_test_subscriber<CallExact>(commit, 1);
-    block_wrapper.subscribe([&test_block](auto block) {
+    block_wrapper.subscribe([test_block](auto block) {
```
Comment (Warchant) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: is it possible to pass by const ref?

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 7]
```diff
@@ -148,30 +136,30 @@ TEST_F(SynchronizerTest, ValidWhenBadStorage) {
...
+  std::shared_ptr<shared_model::interface::Block> test_block =
+      std::make_shared<shared_model::proto::Block>(
+          shared_model::proto::TemplateBlockBuilder<
+              (1 << shared_model::proto::TemplateBlockBuilder<>::total) - 1,
+              shared_model::validation::DefaultBlockValidator>()
+              .height(5)
+              .build()
+              .signAndAddSignature(
+                  shared_model::crypto::DefaultCryptoAlgorithmType::
+                      generateKeypair()));
```
Comment (Warchant) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: same as above: simplify please

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 8]
```diff
@@ -120,7 +107,8 @@ TEST_F(SynchronizerTest, ValidWhenSingleCommitSynchronized) {
 
 TEST_F(SynchronizerTest, ValidWhenBadStorage) {
   // commit from consensus => storage not created => no commit
-  Block test_block;
+  std::shared_ptr<shared_model::interface::Block> test_block =
+      std::make_shared<shared_model::proto::Block>(TestBlockBuilder().build());
```
Comment (Warchant) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: same as above, simplify please

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 9]
```diff
@@ -180,11 +168,11 @@ TEST_F(SynchronizerTest, ValidWhenBlockValidationFailure) {
 
   auto wrapper =
       make_test_subscriber<CallExact>(synchronizer->on_commit_chain(), 1);
-  wrapper.subscribe([&test_block](auto commit) {
+  wrapper.subscribe([test_block](auto commit) {
```
Comment (Warchant) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: const ref?

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 10]
```diff
@@ -180,11 +168,11 @@ TEST_F(SynchronizerTest, ValidWhenBlockValidationFailure) {
...
     auto block_wrapper = make_test_subscriber<CallExact>(commit, 1);
-    block_wrapper.subscribe([&test_block](auto block) {
+    block_wrapper.subscribe([test_block](auto block) {
```
Comment (Warchant) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: const ref?

File: irohad/network/impl/peer_communication_service_impl.cpp [hunk 11]
```diff
@@ -48,8 +48,7 @@ namespace iroha {
       return synchronizer_->on_commit_chain().map([](auto commit) -> Commit {
         return commit.map(
             [](auto block) -> std::shared_ptr<shared_model::interface::Block> {
```
Comment (kamilsa) [inline irohad/network/impl/peer_communication_service_impl.cpp]: Maybe just remove map?

File: irohad/synchronizer/impl/synchronizer_impl.cpp (hunk 3 above)
Comment (kamilsa) [reply irohad/synchronizer/impl/synchronizer_impl.cpp]: again, we should be able to remove map

File: test/module/irohad/synchronizer/synchronizer_test.cpp (hunk 4 above)
Comment (kamilsa) [reply test/module/irohad/synchronizer/synchronizer_test.cpp]: Might not work. Then test block will have type `shared_ptr<shared_model::proto::Block>` whereas `validateBlock` below expects `shared_ptr<shared_model::proto::Block>`. 
If you want to simplify, you need to explicitly define test_block's type

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 12]
```diff
@@ -148,30 +136,30 @@ TEST_F(SynchronizerTest, ValidWhenBadStorage) {
 
 TEST_F(SynchronizerTest, ValidWhenBlockValidationFailure) {
   // commit from consensus => chain validation failed => commit successful
-  Block test_block;
-  test_block.height = 5;
-  test_block.sigs.emplace_back();
-  test_block.hash = iroha::hash(test_block);
+  std::shared_ptr<shared_model::interface::Block> test_block =
+      std::make_shared<shared_model::proto::Block>(
+          shared_model::proto::TemplateBlockBuilder<
```
Comment (kamilsa) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: I understand why do you do that kind of definition of the builder. But better approach is to create TestUnsignedBlockBuilder. Take a look at `test_query_builder.hpp`:
https://github.com/hyperledger/iroha/blob/1adfd054d226c27d3ef0ba341aba6489766a63b1/test/module/shared_model/builders/protobuf/test_query_builder.hpp#L33-L36

File: test/module/irohad/synchronizer/synchronizer_test.cpp (hunk 4 above)
Comment (Warchant) [reply test/module/irohad/synchronizer/synchronizer_test.cpp]: Okay, replace auto with `shared_ptr<shared_model::interface::Block>`, but these pyramides should be simplified.

**Comment by luckychess** [comment]
One comment about const refs: almost all pass by value cases will be changed when other components (like consensus gate) move to a shared model. I do not want to touch them now because all these cases will gone very soon after other components to shared model migration.

File: test/module/irohad/synchronizer/synchronizer_test.cpp (hunk 9 above)
Comment (luckychess) [reply test/module/irohad/synchronizer/synchronizer_test.cpp]: That's just a shared_ptr now so no a big reason to pass it by value.

File: test/module/irohad/synchronizer/synchronizer_test.cpp (hunk 10 above)
Comment (luckychess) [reply test/module/irohad/synchronizer/synchronizer_test.cpp]: Same as above.

File: test/module/irohad/synchronizer/synchronizer_test.cpp (hunk 5 above)
Comment (luckychess) [reply test/module/irohad/synchronizer/synchronizer_test.cpp]: It's not so important for shared pointer.

File: test/module/irohad/synchronizer/synchronizer_test.cpp (hunk 6 above)
Comment (luckychess) [reply test/module/irohad/synchronizer/synchronizer_test.cpp]: See above.

File: test/module/irohad/synchronizer/synchronizer_test.cpp (hunk 5 above)
Comment (muratovv) [reply test/module/irohad/synchronizer/synchronizer_test.cpp]: It is about ownership. Is the code require independent dependency or it used it for some limited time.

File: test/module/shared_model/validators/validators.hpp [hunk 13]
```diff
@@ -47,6 +47,16 @@ namespace shared_model {
       }
     };
 
+    /**
+     * Block validator which always returns answer with no errors.
+     */
+    class BlockAlwaysValidValidator {
```
Comment (muratovv) [inline test/module/shared_model/validators/validators.hpp]: What is the reason to do this class? Maybe rework it with mocks?

File: irohad/synchronizer/impl/synchronizer_impl.cpp [hunk 14]
```diff
@@ -21,7 +21,7 @@
 #include "synchronizer/impl/synchronizer_impl.hpp"
 
 // TODO: 14-02-2018 Alexey Chernyshov remove this after relocation to
-// shared_model https://soramitsu.atlassian.net/browse/IR-903
+// shared_model https://soramitsu.atlassian.net/browse/IR-899
```
Comment (l4l) [inline irohad/synchronizer/impl/synchronizer_impl.cpp]: Is the old model still in use?

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 15]
```diff
@@ -151,31 +161,36 @@ TEST_F(SynchronizerTest, ValidWhenBadStorage) {
...
 TEST_F(SynchronizerTest, ValidWhenBlockValidationFailure) {
   // commit from consensus => chain validation failed => commit successful
-  iroha::model::Block test_block;
-  test_block.height = 5;
-  test_block.sigs.emplace_back();
-  test_block.hash = iroha::hash(test_block);
+  TemplateMockBlockValidator<MockBlockValidator> mockBlockValidator;
+  EXPECT_CALL(*mockBlockValidator.validator, validate(_))
+      .WillOnce(Return(shared_model::validation::Answer()));
+  using TestUnsignedBlockBuilder = shared_model::proto::TemplateBlockBuilder<
```
Comment (l4l) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: Maybe move out to test builders?

Comment (luckychess) [reply test/module/irohad/synchronizer/synchronizer_test.cpp]: It's possible but to do that we also have to move all Mock there what is not a good idea I think.

**Comment by Alexey-N-Chernyshov** [comment]
please pay attention to codacy review https://app.codacy.com/app/stinger112/iroha/pullRequest?prid=1453166

**Pull request merged**